GAME TREE SCENARIOS

sriharikapu opened the pull request:
Covers various scenarios of game tress
Implemented the following codes 

Actions included

- [x] buyer-commit
- [x] buyer-complain
- [x] buyer-redeem
- [x] buyer-refund
- [x] seller-cancel
- [x] seller-commit
- [x] seller-admitFault

Helpers included 

- [x] Delay
- [x] Formatter
- [x]  Check Balances

Also included

- [x] sign Transactions

Tests

- [x] Modularized

Scenarios Covered

- [x] SELLER LISTS A VOUCHER FOR SALE
- [x] BUYER PURCHASES A VOUCHER
- [x] BUYER REDEEMS A COMMITTED VOUCHER
- [x] BUYER CLAIMS REFUND FOR A COMMITTED VOUCHER
- [x] BUYER COMPLAINS ABOUT REDEEMED VOUCHER
- [x] BUYER COMPLAINS ABOUT REFUND VOUCHER
- [x] SELLER ACCEPTS REDEEM FAULT ON COMPLAIN
- [x] SELLER ACCEPTS REFUND FAULT ON COMPLAIN

[Read More](https://github.com/bosonprotocol/contracts/blob/gametree-trace/test/readme.md)

hswopeams reviewed (changes requested):
A couple of things

- It looks like the formatter didn't like something. That's what made the pipeline fail. Try running `./go tests:format_fix` and then recommitting
- Please move the actions folder and everything in it into a folder called testGameTree or similar. Right now it's not clear that the actions folder is for tests
- The tests output using console.log, but I don't see any assertions. It's assertions that make a test a test. For instance when balances are checked in "Check Balances" blocks, there should be an assertion instead of a `console.log`

HristiyanG commented on testGameTree/5-buyer-commit-refund.js:
```diff
@@ -0,0 +1,47 @@
+
+    let voucherSetDetails;
+    let committedVoucher;
+    let refundedVoucher;
```
redundant variable

HristiyanG commented:
Just finished my review, left some comments there, but just still some general questions: 
- Why the old test were deleted? There were roughly about 400 test. Do these scenarios cover each and every missing test?
	- if that is not the case, please bring back the old test
	- please consider writing a script, responsible for the user interactions, and leave the `test` script for the unit testing
- Agreed with Heather that the `actions` folder needs to be moved into the tests, maybe renamed to testGameTreeActions or similar
- Agreed with Heather that the tests should be using assertions
- Do we need to test those actions over a testnet at all?
	- we can do it by running 2 different scripts then, 1 for testnet (e.g. Rinkeby), 1 locally
	- this way while run locally, we should not bother to fill in the config file. We can get the accounts from the local rpc, deploy a contract and fill in the required params automatically
- There should be something at the end indicating that the tests are completed. Currently when the process finishes I was misled that I am doing something wrong as there is not anything which prompts me how tests went
- Maybe a little more detailed README file would be good. I don't think the delay should be that long (30000) if we test the flow locally. Maybe even we could move it into the configs?
- `constant.js` exported from the `test` folder, and now moved to actions/helpers/constants. Do we need to do that. IMHO this should be in the test folder as it's own purpose is to serve the tests
- left some comments for removing the empty lines, commented lines and other

sriharikapu commented on testGameTree/1-seller-create.js:
```diff
@@ -0,0 +1,27 @@
+const {describe,it} = require("mocha");
+let format = require("./helpers/formatter")
+
+
```
removed it

sriharikapu commented on testGameTree/buyer/refundVoucher.js:
```diff
@@ -47,29 +47,32 @@ function refundVoucher(_voucherID) {
+//     await refundVoucher("57896044618658097711785492504343954004559654357715190152841577105831485243398");
+// })();
 
-// module.exports = RefundVoucherETHETH;
```
removed them

sriharikapu commented on testGameTree/buyer/refundVoucher.js:
```diff
@@ -47,29 +47,32 @@ function refundVoucher(_voucherID) {
 
-                console.log(output)
-                return(output)
+                // console.log(output)
```
removed

sriharikapu commented on testGameTree/helpers/checkBalance.js:
```diff
@@ -0,0 +1,34 @@
+}
+
+module.exports = checkBalances;
+
+
```
resolved

sriharikapu commented on testGameTree/helpers/formatter.js:
```diff
@@ -0,0 +1,19 @@
+let Table = require('cli-table');
+
+async function formatter(_input) {
+    // console.log(_input);
```
removed the commented line

sriharikapu commented on testGameTree/helpers/formatter.js:
```diff
@@ -0,0 +1,19 @@
+}
+
+module.exports = formatter;
+
+
```
removed the commented lines

sriharikapu commented on testGameTree/seller/cancelVoucherSet.js:
```diff
@@ -17,11 +17,11 @@ Contract.setProvider(PROVIDER);
 // const seller = new ethers.Wallet(SELLER_SECRET, PROVIDER);
```
removed

sriharikapu commented on testGameTree/seller/cancelVoucherSet.js:
```diff
@@ -17,11 +17,11 @@ Contract.setProvider(PROVIDER);
 // const seller = new ethers.Wallet(SELLER_SECRET, PROVIDER);
 const seller = SELLER_PUBLIC;
 
-// let sellerRequestCancelorFault =
```
removed

sriharikapu commented on testGameTree/seller/cancelVoucherSet.js:
```diff
@@ -49,36 +49,42 @@ const seller = SELLER_PUBLIC;
-                console.log("Transaction Hash : "+hash);
+                console.log("Transaction Hash : ",hash);
             }).on('receipt', function(receipt){
                 // console.log(receipt)
```
removed

sriharikapu commented on testGameTree/seller/cancelVoucherSet.js:
```diff
@@ -49,36 +49,42 @@ const seller = SELLER_PUBLIC;
 
-                console.log(output)
-                return(output)
+                // console.log(output)
```
removed

sriharikapu commented on testGameTree/seller/cancelVoucherSet.js:
```diff
@@ -49,36 +49,42 @@ const seller = SELLER_PUBLIC;
+//     await requestCancelorFault("57896044618658097711785492504343954004219371990794251689378202498399717031936");
+// })();
 
-// module.exports = sellerRequestCancelorFault;
```
removed

sriharikapu commented on testGameTree/seller/createVoucher.js:
```diff
@@ -0,0 +1,103 @@
+const helpers = require('../helpers/constants')
+const Tx = require('ethereumjs-tx').Transaction;
+let converter = require('hex2dec');
+
```
removed

sriharikapu commented on testGameTree/seller/createVoucher.js:
```diff
@@ -0,0 +1,103 @@
+// set provider for all later instances to use
+Contract.setProvider(PROVIDER);
+
+// const seller = new ethers.Wallet(SELLER_SECRET, PROVIDER);
```
removed

sriharikapu commented on testGameTree/seller/createVoucher.js:
```diff
@@ -0,0 +1,103 @@
+                }
+
+                console.log("Transaction Hash : ",hash);
+                // resolve(hash)
```
removed

sriharikapu commented on testGameTree/seller/createVoucher.js:
```diff
@@ -0,0 +1,103 @@
+            }).on('receipt', function(receipt){
+                // console.log(receipt);
+                // console.log(receipt.topics[1]);
+                // console.log(receipt.topics[2]);
```
removed

sriharikapu commented on testGameTree/seller/createVoucher.js:
```diff
@@ -0,0 +1,103 @@
+                    "logReceipt3": receipt.logs[2].id
+                }
+
+                // console.log(output)
```
removed

sriharikapu commented on testGameTree/seller/createVoucher.js:
```diff
@@ -0,0 +1,103 @@
+// (async function newOrder () {
+//     await CreateOrderETHETH();
+// })();
+
```
removed

sriharikapu commented on testGameTree/seller/faultVoucher.js:
```diff
@@ -45,33 +44,37 @@ function faultVoucher(_voucherSetID) {
             web3.eth.sendSignedTransaction('0x' + serializedTx.toString('hex'), (err, hash) => {
                 if(err) {
-                    console.log(err)
+                    // console.log(err)
```
removed

sriharikapu commented on testGameTree/seller/faultVoucher.js:
```diff
@@ -45,33 +44,37 @@ function faultVoucher(_voucherSetID) {
 
-                resolve(hash)
-            }).on('transactionHash', function(hash){
+                // resolve(hash)
```
removed

sriharikapu commented on testGameTree/seller/faultVoucher.js:
```diff
@@ -45,33 +44,37 @@ function faultVoucher(_voucherSetID) {
 
-                console.log(output)
-                return(output)
+                // console.log(output)
```
removed

sriharikapu commented on testGameTree/seller/faultVoucher.js:
```diff
@@ -45,33 +44,37 @@ function faultVoucher(_voucherSetID) {
-
+// (async function newOrder () {
+//     await faultVoucher("57896044618658097711785492504343954004219371990794251689378202498399717031936");
+// })();
```
removed

sriharikapu commented on testGameTree/seller/faultVoucher.js:
```diff
@@ -45,33 +44,37 @@ function faultVoucher(_voucherSetID) {
+                // resolve(hash)
                 console.log("Transaction Hash : "+hash);
             }).on('receipt', function(receipt){
                 // console.log(receipt)
```
removed

sriharikapu commented on testGameTree/2-seller-create-cancel.js:
```diff
@@ -0,0 +1,34 @@
+let format = require("./helpers/formatter");
+const checkBalance = require("./helpers/checkBalance");
+
+
```
removed

sriharikapu commented on testGameTree/4-buyer-commit-redeem.js:
```diff
@@ -0,0 +1,48 @@
+const {describe,it} = require("mocha");
+
+let format = require("./helpers/formatter")
+
```
removed

sriharikapu commented on testGameTree/6-buyer-commit-redeem-complain.js:
```diff
@@ -0,0 +1,56 @@
+const {describe,it} = require("mocha");
+
+let format = require("./helpers/formatter");
+
```
removed

sriharikapu commented on testGameTree/7-buyer-commit-refund-complain.js:
```diff
@@ -0,0 +1,55 @@
+        console.log(await checkBalance());
+        let refund = await refundVoucher(committedVoucher["MintedVoucherID"]);
+        await format(refund);
+
```
removed

sriharikapu commented on testGameTree/8-buyer-commit-redeem-complain-seller-fault.js:
```diff
@@ -0,0 +1,65 @@
+const TIMEOUT = 500 * 1000;
+
+describe("TEST SCENARIO 08 :: SELLER CREATES, BUYER COMMITS, REDEEMS & COMPLAINS, SELLER FAULTS", async function() {
+
```
removed

sriharikapu commented on testGameTree/9-buyer-commit-refund-complain-seller-fault.js:
```diff
@@ -0,0 +1,63 @@
+const {describe,it} = require("mocha");
+let format = require("./helpers/formatter");
+
+
```
removed

sriharikapu commented on testGameTree/9-buyer-commit-seller-fault.js:
```diff
@@ -0,0 +1,42 @@
+const sellerCreate = require("./seller/createVoucher");
+// const sellerCancel = require("./seller/cancelVoucherSet");
```
removed

sriharikapu commented on testGameTree/9-buyer-commit-seller-fault.js:
```diff
@@ -0,0 +1,42 @@
+const checkBalance = require("./helpers/checkBalance");
+const {describe,it} = require("mocha");
+
+
```
removed

sriharikapu commented on testGameTree/9-buyer-commit-seller-fault.js:
```diff
@@ -0,0 +1,42 @@
+const TIMEOUT = 500 * 1000;
+
+describe("TEST SCENARIO 01 :: SELLER CREATE", async function() {
+
```
removed

sriharikapu commented on testGameTree/buyer/commitVoucher.js:
```diff
@@ -50,13 +49,14 @@ const buyer = BUYER_PUBLIC;
+                // resolve(hash)
                 console.log("Transaction Hash : "+hash);
             }).on('receipt', function(receipt){
                 // console.log(receipt);
```
removed

sriharikapu commented on testGameTree/buyer/commitVoucher.js:
```diff
@@ -65,29 +65,32 @@ const buyer = BUYER_PUBLIC;
                     "promiseID":promiseID,
+                    "gasPaid":converter.hexToDec(gasPaid),
+                    "gasUsed":gasUsed,
                     "correlationID":correlationID,
```
mofied

sriharikapu commented on testGameTree/buyer/compainVoucher.js:
```diff
@@ -16,20 +16,19 @@ Contract.setProvider(PROVIDER);
 
 const buyer = BUYER_PUBLIC;
 
-// let ComplainVoucherETHETH =
```
removed

sriharikapu commented on testGameTree/buyer/compainVoucher.js:
```diff
@@ -47,29 +46,30 @@ function complainVoucher(_voucherID) {
-            }).on('transactionHash', function(hash){
                 console.log("Transaction Hash : "+hash);
             }).on('receipt', function(receipt){
                 // console.log(receipt);
```
removed

sriharikapu commented on testGameTree/buyer/compainVoucher.js:
```diff
@@ -47,29 +46,30 @@ function complainVoucher(_voucherID) {
 
-                console.log(output)
-                return(output)
+                // console.log(output)
```
removed

sriharikapu commented on testGameTree/buyer/compainVoucher.js:
```diff
@@ -47,29 +46,30 @@ function complainVoucher(_voucherID) {
+//     await complainVoucher("57896044618658097711785492504343954004559654357715190152841577105831485243399");
+// })();
 
-// module.exports = ComplainVoucherETHETH;
```
removed

sriharikapu commented on testGameTree/buyer/redeemVoucher.js:
```diff
@@ -1,13 +1,13 @@
 let Web3 = require('web3');
-const BN = require('bn.js');
+// const BN = require('bn.js');
```
removed

sriharikapu commented on testGameTree/buyer/redeemVoucher.js:
```diff
@@ -1,13 +1,13 @@
+// const BN = require('bn.js');
 let Contract = require('web3-eth-contract');
-const helpers = require('../helpers/constants')
+// const helpers = require('../helpers/constants')
```
removed

sriharikapu commented on testGameTree/buyer/redeemVoucher.js:
```diff
@@ -1,13 +1,13 @@
+// const helpers = require('../helpers/constants')
 const Tx = require('ethereumjs-tx').Transaction;
 let converter = require('hex2dec');
 
```
removed

sriharikapu commented on testGameTree/buyer/redeemVoucher.js:
```diff
@@ -16,20 +16,19 @@ Contract.setProvider(PROVIDER);
 
 const buyer = BUYER_PUBLIC;
 
-// let RedeemVoucherETHETH =
```
removed

sriharikapu commented on testGameTree/buyer/redeemVoucher.js:
```diff
@@ -47,34 +46,36 @@ const buyer = BUYER_PUBLIC;
+                // resolve(hash)
                 console.log("Transaction Hash : "+hash);
             }).on('receipt', function(receipt){
                 // console.log(receipt);
```
removed

sriharikapu commented on testGameTree/buyer/redeemVoucher.js:
```diff
@@ -47,34 +46,36 @@ const buyer = BUYER_PUBLIC;
 
-                console.log(output)
-                return(output)
+                // console.log(output)
```
removed

sriharikapu commented on testGameTree/buyer/redeemVoucher.js:
```diff
@@ -47,34 +46,36 @@ const buyer = BUYER_PUBLIC;
+//     await redeemVoucher("57896044618658097711785492504343954004559654357715190152841577105831485243399");
+// })();
 
-// module.exports = RedeemVoucherETHETH;
```
removed

sriharikapu commented on testGameTree/buyer/refundVoucher.js:
```diff
@@ -47,29 +47,32 @@ function refundVoucher(_voucherID) {
-                resolve(hash)
-            }).on('transactionHash', function(hash){
+                // resolve(hash)
+                // console.log(hash);
```
removed

sriharikapu commented:
Why the old test were deleted? There were roughly about 400 tests. Do these scenarios cover each and every missing test?
 - [x] Brought Back the previous test
 
 Agreed with Heather that the actions folder needs to be moved into the tests, maybe renamed to testGameTreeActions or similar
 - [x] Renamed to testGameTree
 
 Agreed with Heather that the tests should be using assertions
 - [x] Added More assertions to tests
 
Do we need to test those actions over a testnet at all?
 - [x] You can set the provider to local or any testnet url so you can run them
 
There should be something at the end indicating that the tests are completed. Currently when the process finishes I was misled that I am doing something wrong as there is not anything which prompts me how tests went
- [x] once the tests finish you will be able to see the number of tests passing, 
 
Maybe a little more detailed README file would be good. I don't think the delay should be that long (30000) if we test the flow locally. Maybe even we could move it into the configs?
- [x] delay was set to 30000ms initially because of the delay in transactions getting confirmed lately, I had kept the delay to the config so you can customized for local test and testnet etc

constant.js exported from the test folder, and now moved to actions/helpers/constants. Do we need to do that. IMHO this should be in the test folder as it's own purpose is to serve the tests
- [x] Actually we don't need all the variables in the constants from the `constants.js`, I had removed the unused variables. also we can link to the previous constants

left some comments for removing the empty lines, commented lines and other
- [x] removed unnecessary empty lines mentioned

pipeline should succeed before merging
- [x]  made sure that pipeline passes

sriharikapu commented:
@hswopeams I think we can merge this once @HristiyanG approves it. Also I checked at my end everything works as expected.

hswopeams commented:
Changed gametree tests to run locally on Ganace instance started by the (Ruby) `go` script. Discussed with Holly, and it was decided that running locally is enough. There's no need to add complexity by adding config to also run against a testnet.

thecryptofruit commented:
@damir-bajramovic , please have a look if this is in your card, otherwise close. thanks

mischat commented:
All of the Game Tree has been incorporated into our unit test suite, we are happy that we are covering all of the game tree and more in `main` of this repo. This branch will be deleted.
